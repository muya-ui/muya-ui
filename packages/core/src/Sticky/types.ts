import React from 'react';
import { ICustomStyleBaseProps } from '../types';

/**
 * Step style keys
 * @styles sticky 固定节点
 */
export type IStickyStyleKeys = 'sticky';

export interface IStickyPureProps extends ICustomStyleBaseProps<IStickyStyleKeys> {
  /**
   * 设置需要监听其滚动事件的元素，值为一个返回对应 DOM 元素的函数
   *
   * @default () => window
   * @memberof IStickyPureProps
   */
  target?: () => HTMLElement;
  /**
   * 距离窗口顶部达到指定偏移量后触发
   *
   * @memberof IStickyPureProps
   */
  offsetTop?: number;
  /**
   * 固定状态改变时触发的回调函数
   *
   * @memberof IStickyPureProps
   */
  onStatusChange?: (fixed: boolean) => void;
}

export type IStickyProps = IStickyPureProps & Omit<React.HTMLAttributes<HTMLDivElement>, 'target'>;
export type ScrollElement = HTMLElement | Window;
