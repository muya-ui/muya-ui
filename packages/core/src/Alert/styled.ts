import styled, { css } from 'styled-components';

import { IComponentSizeSpec } from '@muya-ui/theme-light';

import { StyledButton, StyledLink } from '../Button/styled';
import { IThemedBaseProps } from '../types';
import { IAlertProps } from './types';

interface IBaseSizeProps {
  $size: IComponentSizeSpec;
}

export interface IStyledAlertProps extends IThemedBaseProps, IBaseSizeProps {
  $type: Required<IAlertProps>['type'];
  $hasTitle?: boolean;
  $hasDesc?: boolean;
  $showIcon?: boolean;
  $showClose?: boolean;
}

export interface IStyledIconProps extends IThemedBaseProps, IBaseSizeProps {
  $type: Required<IAlertProps>['type'];
  $hasTitleAndDesc?: boolean;
}

export interface IStyledCloseProps extends IThemedBaseProps, IBaseSizeProps {
  $hasTitleAndDesc?: boolean;
}

export interface IStyledTitleProps extends IThemedBaseProps, IBaseSizeProps {
  $hasDesc?: boolean;
}

export const StyledAlert = styled.div`
  ${(props: IStyledAlertProps) => {
    const { $type, $hasTitle, $hasDesc, $showIcon, $showClose, $size, theme } = props;
    const {
      typeBg,
      typeBorder,
      borderRadius,
      padding,
      hasIconPaddingLeft,
      hasIconAndDescPaddingLeft,
      hasCloseIconPaddingRight,
      hasTitleAndDescPaddingVertical,
    } = theme.components.Alert;
    let [top, right, bottom, left] = padding[$size];
    if ($showIcon) {
      left = $hasDesc ? hasIconAndDescPaddingLeft[$size] : hasIconPaddingLeft[$size];
    }
    if ($showClose) {
      right = hasCloseIconPaddingRight[$size];
    }
    if ($hasTitle && $hasDesc) {
      top = hasTitleAndDescPaddingVertical[$size];
      bottom = hasTitleAndDescPaddingVertical[$size];
    }
    return css`
      position: relative;
      box-sizing: border-box;
      background-color: ${typeBg[$type!]};
      border: ${typeBorder[$type!]};
      border-radius: ${borderRadius};

      /** 直接改 Button 存在风险，因此覆盖 */
      & ${StyledButton}, & ${StyledLink} {
        vertical-align: unset;
      }

      ${StyledAlertContainer} {
        padding: ${top}px ${right}px ${bottom}px ${left}px;
      }
    `;
  }}
`;

export const StyledAlertContainer = styled.div`
  position: relative;
  margin: 0 auto;
`;

export const StyledIcon = styled.div`
  ${(props: IStyledIconProps) => {
    const { $type, theme, $hasTitleAndDesc, $size } = props;
    const {
      left: iconLeft,
      hasTitleAndDescLeft,
      top: iconTop,
      hasTitleAndDescTop,
    } = theme.components.Alert.icon;
    const { lineHeight } = theme.components.Alert.title;
    const finalLineHeight = $hasTitleAndDesc ? 1 : `${lineHeight[$size]}px`;
    const left = $hasTitleAndDesc ? hasTitleAndDescLeft : iconLeft[$size];
    const top = $hasTitleAndDesc ? hasTitleAndDescTop : iconTop[$size];
    return css`
      position: absolute;
      left: ${left}px;
      top: ${top}px;
      color: ${theme.colors.pattern.feature[$type!]};
      font-size: 0;
      line-height: ${finalLineHeight};
    `;
  }}
`;

export const StyledClose = styled.div`
  ${(props: IStyledCloseProps) => {
    const { theme, $hasTitleAndDesc, $size } = props;
    const { closeIconTop, closeIconRight, hasTitleAndDescCloseIconTop } = theme.components.Alert;
    let finalTop = closeIconTop[$size];
    if ($hasTitleAndDesc) {
      finalTop = hasTitleAndDescCloseIconTop;
    }
    return css`
      position: absolute;
      right: ${closeIconRight}px;
      top: ${finalTop}px;
      line-height: 0;
    `;
  }}
`;

export const StyledTitle = styled.div`
  ${(props: IStyledTitleProps) => {
    const { theme, $size, $hasDesc } = props;
    const {
      colors,
      typography: {
        spec: { fontWeight },
      },
    } = theme;
    const {
      fontSize,
      lineHeight,
      textFine,
      hasDescTextFine,
      hasDescFontSize,
      hasDescLineHeight,
      hasDescMarginBottom,
    } = theme.components.Alert.title;
    const finalTextFine = $hasDesc ? hasDescTextFine : textFine;
    const finalFontSize = $hasDesc ? hasDescFontSize : fontSize[$size];
    const finalLineHeight = $hasDesc ? hasDescLineHeight : lineHeight[$size];
    const finalMarginBottom = $hasDesc ? hasDescMarginBottom : 0;
    return css`
      color: ${colors.pattern.text.title};
      font-size: ${finalFontSize}px;
      line-height: ${finalLineHeight}px;
      font-weight: ${finalTextFine ? fontWeight.regular : fontWeight.semibold};
      margin-bottom: ${finalMarginBottom}px;
    `;
  }}
`;

export const StyledDesc = styled.span`
  ${(props: IThemedBaseProps) => {
    const { theme } = props;
    const { colors } = theme;
    const { fontSize, lineHeight } = theme.components.Alert.desc;
    return css`
      color: ${colors.pattern.text.text};
      font-size: ${fontSize}px;
      line-height: ${lineHeight}px;
    `;
  }}
`;
