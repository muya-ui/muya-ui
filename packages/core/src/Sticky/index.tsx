export { default } from './Sticky';
export { default as Sticky } from './Sticky';
export * from './types';
