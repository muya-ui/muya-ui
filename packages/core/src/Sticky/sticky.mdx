---
name: Sticky 粘性布局
menu: 组件
route: /components/sticky
autoDemos: true
parent: 粘性布局
disableDoubleRow: true
---

# Sticky 粘性布局

将页面元素钉在容器范围内。

## 使用场景

当内容区域比较长，需要滚动页面时，这部分内容对应的操作或者导航需要在滚动范围内始终展现。常用于侧边菜单和按钮组合。

页面可视范围过小时，慎用此功能以免遮挡页面内容。

{Demos}

### API

#### `IStickyProps`

<Props propsName="IStickyProps" />

### Styles

#### `IStickyStyleKeys`

<Styles typeName="IStickyStyleKeys" componentName="sticky" />