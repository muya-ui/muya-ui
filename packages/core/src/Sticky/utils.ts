import { ScrollElement } from './types';
const overflowScrollReg = /scroll|auto/i;

function isWindow(val: unknown): val is Window {
  return val === window;
}

export function getScroller(el: HTMLElement, root: ScrollElement = window) {
  let node = el;

  while (node && node.tagName !== 'HTML' && node.nodeType === 1 && node !== root) {
    const { overflowY } = window.getComputedStyle(node);

    if (overflowY && overflowScrollReg.test(overflowY)) {
      if (node.tagName !== 'BODY') {
        return node;
      }

      const { overflowY: htmlOverflowY } = window.getComputedStyle(node.parentNode as Element);

      if (htmlOverflowY && overflowScrollReg.test(htmlOverflowY)) {
        return node;
      }
    }
    node = node.parentNode as HTMLElement;
  }

  return root;
}

export function isHidden(el: HTMLElement) {
  const style = window.getComputedStyle(el);
  const hidden = style.display === 'none';

  const parentHidden = el.offsetParent === null && style.position !== 'fixed';

  return hidden || parentHidden;
}

export function getScrollTop(el: ScrollElement): number {
  return 'scrollTop' in el ? el.scrollTop : el.pageYOffset;
}

export function getRootScrollTop(): number {
  return window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop || 0;
}

export function getElementTop(el: ScrollElement, scroller?: HTMLElement) {
  if (isWindow(el)) {
    return 0;
  }
  const scrollTop = scroller ? getScrollTop(scroller) : getRootScrollTop();
  return el.getBoundingClientRect().top + scrollTop;
}
