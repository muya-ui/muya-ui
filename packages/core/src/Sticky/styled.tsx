import styled, { css } from 'styled-components';

interface IStyledStickyProps {
  $height?: number | null;
  $fixed?: boolean;
  $offsetTop?: number;
  $transform?: number;
}

export const StyledContainer = styled.div<IStyledStickyProps>`
  ${props => {
    const { $height } = props;
    if ($height) {
      return css`
        height: ${$height}px;
      `;
    }
  }}
`;

export const StyledSticky = styled.div<IStyledStickyProps>`
  ${props => {
    if (props.$fixed) {
      return css`
        position: fixed;
        top: 0;
        z-index: 99;
      `;
    }
  }}
`;
