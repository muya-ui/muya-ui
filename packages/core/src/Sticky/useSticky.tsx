import { useState, useCallback, useMemo, CSSProperties, RefObject } from 'react';
import { getScroller, isHidden, getScrollTop, getElementTop } from './utils';
import { IStickyProps, ScrollElement } from './types';
import { useEffectOnce, useEventCallback } from '@muya-ui/utils';

export default function useSticky(props: IStickyProps, divEl: RefObject<HTMLDivElement>) {
  const { offsetTop, target, onStatusChange } = props;
  const [fixed, setFixed] = useState(false);
  const [transform, setTransform] = useState(0);
  const [scrollTick, setScrollTick] = useState(false);
  const [targetScrollTick, setTargetScrollTick] = useState(false);

  const handleChange = useCallback(
    (status: boolean) => {
      if (onStatusChange && fixed !== status) {
        onStatusChange(status);
      }
    },
    [fixed, onStatusChange],
  );

  const calculate = useCallback(
    (top: number, target: HTMLElement, current: HTMLElement) => {
      const transform = target.getBoundingClientRect().top;
      const distance = transform - current.getBoundingClientRect().top;
      const originalPosition = Math.abs(distance - getScrollTop(target));
      const scrollTop = getScrollTop(target);
      if (scrollTop === 0) {
        setFixed(false);
        handleChange(false);
        return;
      }
      if (originalPosition - scrollTop <= top) {
        setTransform(transform);
      } else {
        setTransform(transform + originalPosition - scrollTop - top);
      }
      setFixed(true);
      handleChange(true);
    },
    [handleChange],
  );

  const onScroll = useEventCallback(() => {
    setScrollTick(false);
    const current = divEl.current;
    if (!current) return;
    if (isHidden(current)) {
      return;
    }
    const top = Number(offsetTop);
    let scrollTop = getScrollTop(window);
    const topToPageTop = getElementTop(current);
    if (target) {
      if (!fixed) {
        return;
      }
      calculate(top, target(), current);
      return;
    }
    if (scrollTop + top > topToPageTop) {
      setFixed(true);
      handleChange(true);
      setTransform(0);
    } else {
      setFixed(false);
      handleChange(false);
    }
  }, [divEl, target, fixed, offsetTop]);

  const handleScroll = useCallback(() => {
    !scrollTick && requestAnimationFrame(onScroll);
    setScrollTick(true);
  }, [onScroll, scrollTick]);

  const onTargetScroll = useCallback(() => {
    setTargetScrollTick(false);
    const current = divEl.current;
    if (!current || !target) return;
    if (isHidden(current)) {
      return;
    }

    const top = Number(offsetTop);
    calculate(top, target(), current);
  }, [divEl, target, offsetTop, calculate]);

  const handleTargetScroll = useCallback(() => {
    !targetScrollTick && requestAnimationFrame(onTargetScroll);
    setTargetScrollTick(true);
  }, [onTargetScroll, targetScrollTick]);

  useEffectOnce(() => {
    const current = divEl.current;
    let scroller: ScrollElement = window;

    if (current) {
      scroller = getScroller(current);
      if (target) {
        window.addEventListener('scroll', handleScroll);
        target().addEventListener('scroll', handleTargetScroll, true);
      } else {
        scroller.addEventListener('scroll', handleScroll);
      }
    }

    return () => {
      if (target) {
        window.removeEventListener('scroll', handleScroll);
        target().removeEventListener('scroll', handleTargetScroll, true);
      }
      scroller.removeEventListener('scroll', handleScroll);
    };
  });

  const style = useMemo(() => {
    if (!fixed) {
      return;
    }
    const style: CSSProperties = {};
    if (Number(offsetTop)) {
      style.top = `${offsetTop}px`;
    }
    if (transform) {
      style.transform = `translate3d(0, ${transform}px, 0)`;
      style.WebkitTransform = `translate3d(0, ${transform}px, 0)`;
      style.willChange = 'transform';
    }

    return style;
  }, [fixed, offsetTop, transform]);

  return {
    style,
    fixed,
    divEl,
  };
}
