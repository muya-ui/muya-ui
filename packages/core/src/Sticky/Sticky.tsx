import React, { useRef, Ref } from 'react';
import { useForkRef } from '@muya-ui/utils';
import memoForwardRef from '../utils/memoForwardRef';
import { IStickyProps } from './types';
import { StyledContainer, StyledSticky } from './styled';
import useStyles from '../utils/useStyles';
import useSticky from './useSticky';

const defaultStyles = {
  sticky: '',
};

export default memoForwardRef((props: IStickyProps, ref: Ref<HTMLDivElement>) => {
  const { children, styles, target, onStatusChange, ...others } = props;
  const innerStyles = useStyles('sticky', defaultStyles, styles);
  const divEl = useRef<HTMLDivElement>(null);
  const handleRef = useForkRef(ref, divEl);
  const { style, fixed } = useSticky(props, divEl);
  const height = (divEl.current && divEl.current.offsetHeight) || 0;
  const innerStyle = fixed ? style : {};

  return (
    <StyledContainer $height={fixed ? height : null} ref={handleRef} {...others}>
      <StyledSticky
        $fixed={fixed}
        {...innerStyles.sticky}
        style={{ ...innerStyles.sticky.style, ...innerStyle }}
      >
        {children}
      </StyledSticky>
    </StyledContainer>
  );
});
