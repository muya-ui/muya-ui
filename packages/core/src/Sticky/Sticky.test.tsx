import React, { createRef } from 'react';
import renderer from 'react-test-renderer';

import Sticky from './Sticky';
import { act, renderHook } from '@testing-library/react-hooks';
import useSticky from './useSticky';
import { getScroller, getElementTop } from './utils';
import { StyledContainer, StyledSticky } from './styled';

const getDivAndRef = () => {
  const ref = createRef<HTMLDivElement>();
  (ref as any).current = {
    getBoundingClientRect: jest.fn(() => ({ top: 10 })),
  };
  const div = document.createElement('div');
  document.body.appendChild(div);
  return {
    div,
    ref,
  };
};

beforeAll(() => {
  // @ts-ignore
  window.getComputedStyle = jest.fn(x => ({
    overflowY: 'scroll',
  }));
  // @ts-ignore
  jest.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => cb());
  Object.defineProperty(HTMLElement.prototype, 'offsetParent', {
    get() {
      return this.parentNode;
    },
  });
});

test('测试 Sticky 正常渲染', () => {
  const tree = renderer.create(<Sticky>测试</Sticky>).toJSON();
  act(() => {
    window.dispatchEvent(new CustomEvent('scroll', { detail: 10 }));
  });
  expect(tree).toMatchSnapshot();
});

test('测试 window 滚动', async () => {
  const ref = createRef<HTMLDivElement>();
  (ref as any).current = {
    getBoundingClientRect: jest.fn(() => ({ top: 10 })),
    offsetHeight: 10,
  };
  const { result } = renderHook(() =>
    useSticky(
      {
        offsetTop: 20,
      },
      ref,
    ),
  );
  act(() => {
    window.dispatchEvent(new CustomEvent('scroll', { detail: 10 }));
  });
  expect(result.current.style!.top).toBe('20px');
  expect(result.current.fixed).toBe(true);
  act(() => {
    (ref as any).current = {
      getBoundingClientRect: jest.fn(() => ({ top: 30 })),
    };
    window.dispatchEvent(new CustomEvent('scroll', { detail: 10 }));
  });
  expect(result.current.fixed).toBe(false);
});

test('测试 target 滚动', async () => {
  const { div, ref } = getDivAndRef();
  div.scrollTop = 20;
  const { result } = renderHook(() =>
    useSticky(
      {
        offsetTop: 20,
        target: () => div,
      },
      ref,
    ),
  );
  act(() => {
    div.dispatchEvent(new CustomEvent('scroll', { detail: 10 }));
  });
  expect(result.current.style!.top).toBe('20px');
  expect(result.current.fixed).toBe(true);
  act(() => {
    (ref as any).current = {
      getBoundingClientRect: jest.fn(() => ({ top: 40 })),
    };
    div.dispatchEvent(new CustomEvent('scroll', { detail: 40 }));
  });
  expect(result.current.style!.transform).toBe('translate3d(0, 20px, 0)');

  const { div: div1 } = getDivAndRef();
  div1.scrollTop = 0;
  const { unmount } = renderHook(() =>
    useSticky(
      {
        offsetTop: 20,
        target: () => div1,
      },
      ref,
    ),
  );
  act(() => {
    div1.dispatchEvent(new CustomEvent('scroll', { detail: 10 }));
    window.dispatchEvent(new CustomEvent('scroll', { detail: 10 }));
  });
  unmount();
});

test('getScroller', () => {
  const div = document.createElement('div');
  document.body.appendChild(div);
  expect(getScroller(div)).toBeInstanceOf(HTMLDivElement);
  expect(getScroller(document.body)).toBeInstanceOf(HTMLBodyElement);
});

test('getElementTop', () => {
  expect(getElementTop(window)).toBe(0);
  const div = document.createElement('div');
  expect(getElementTop(div, document.body)).toBe(0);
});

test('Styled', () => {
  const wrapper1 = renderer.create(<StyledContainer $height={10} />).toJSON();
  expect(wrapper1).toMatchSnapshot();
  const wrapper2 = renderer.create(<StyledSticky $fixed />).toJSON();
  expect(wrapper2).toMatchSnapshot();
});

test('current undefined', async () => {
  const ref = createRef<HTMLDivElement>();
  (ref as any).current = null;
  const { result } = renderHook(() =>
    useSticky(
      {
        offsetTop: 20,
      },
      ref,
    ),
  );
  const div = document.createElement('div');
  document.body.appendChild(div);
  const { result: result1 } = renderHook(() =>
    useSticky(
      {
        offsetTop: 20,
        target: () => div,
      },
      ref,
    ),
  );
  expect(result.current.fixed).toBe(false);
  expect(result1.current.fixed).toBe(false);
});
