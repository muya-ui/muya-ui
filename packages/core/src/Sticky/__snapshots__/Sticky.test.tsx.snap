// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styled 1`] = `
.c0 {
  height: 10px;
}

<div
  className="c0"
/>
`;

exports[`Styled 2`] = `
.c0 {
  position: fixed;
  top: 0;
  z-index: 99;
}

<div
  className="c0"
/>
`;

exports[`测试 Sticky 正常渲染 1`] = `
<div
  className=""
>
  <div
    className="muya-sticky-sticky"
    style={Object {}}
  >
    测试
  </div>
</div>
`;
